### ISSUE TITLE
Support `footnoteAnchorPrefix` for `goldmark`

### ISSUE THREAD

**Issue by shaform** [gohugoio/hugo#8045]
It looks like goldmark has implemented the `footnoteAnchorPrefix` feature https://github.com/yuin/goldmark/commit/9e0189df270cb443c204b51e89e2b0749b98370c. It would be great if hugo can also support this option.

Also see discussions in [Goldmark: How to configure the footnote anchor prefix?
](https://discourse.gohugo.io/t/goldmark-how-to-configure-the-footnote-anchor-prefix/29639)

**Comment by xpufx** [comment]
Anybody looking into this? Looking at the date of the report it might even be fixed already but upgrading Goldmark but it's not clear. No sign of it in documentation though

**Comment by shaform** [comment]
Yeah, it would be nice if this feature could be implemented now that blackfriday can no longer be used.

**Comment by clsty** [comment]
Any luck on this please? I've just met the footnote Id clash thing with multiple posts on one single page.

**Comment by shaform** [comment]
> Any luck on this please? I've just met the footnote Id duplicating with multiple posts on one single page.

@clsty Do you know if this problem is still present? I haven't had time to test it myself recently.

**Comment by clsty** [comment]
> @clsty Do you know if this problem is still present? I haven't had time to test it myself recently.

Sorry but I'm also busy recently.

**Comment by jmooring** [comment]
There are at least two use cases when rendering a list page, and each use case requires a different `href` attribute.

### Example 1

```text
{{ range site.RegularPages.ByTitle }}
  <h2><a href="{{ .RelPermalink }}">{{ .LinkTitle }}</a></h2>
  {{ .Content }}
{{ end }}
```

<img width="750" height="493" alt="Image" src="https://github.com/user-attachments/assets/aa1a4e0f-4345-4dae-84b8-ad09ffef6439" />

In the above, the `href` attribute must point to a fragment on the current (list) page.

### Example 2

```text
{{ range site.RegularPages.ByTitle }}
  <h2><a href="{{ .RelPermalink }}">{{ .LinkTitle }}</a></h2>
  {{ .Summary }}
{{ end }}
```

<img width="745" height="250" alt="Image" src="https://github.com/user-attachments/assets/a574822b-1660-460f-adfb-da9c17d67b83" />

In the above, the `href` attribute must point to a fragment on the source (single) page. 

Also, in both examples, the footnote number is the same for both list items, which doesn't make a lot of sense to me. Bottom line... the objective is unclear to me.

**Comment by jmooring** [comment]
Another approach, if you want to render the content and not just the summary, is to create a list shortcode.

### Example 3

```text
{{ range site.RegularPages.ByTitle }}
## [{{ .LinkTitle }}]({{ .RelPermalink }})
{{ .RenderShortcodes }}{{/* Do not indent. */}}
{{ end }}
```

And call it like this:

```text
{{% list %}}
```

That yields something like this:

<img width="741" height="427" alt="Image" src="https://github.com/user-attachments/assets/87b6f6ac-7f4e-4f45-9015-ed800d2599a4" />

But you get the wrong result because both source pages use the same footnote number.

---

This the test site I used to generate the examples above:

```text
git clone --single-branch -b hugo-github-issue-8045 https://github.com/jmooring/hugo-testing hugo-github-issue-8045
cd hugo-github-issue-8045
hugo server
```

**Comment by shaform** [comment]
@jmooring In my case, I wanted to render the full content, including the footnotes, and allow the links to correctly link to the corresponding articles in the same list page. It used to be possible to achieve this with `footnoteAnchorPrefix for Blackfriday`. The idea is to choose a different prefix for each article, then it won't conflict with each other even if their footnotes have the same numbers. However, this approach stops working after we switch to `goldmark`.

One way to fix the above problem is to also support footnoteAnchorPrefix for `goldmark`, then we could choose different prefixes for different articles again, but there could be other solutions as long as we could make those footnote links work correctly.

**Comment by shaform** [comment]
Wondering if it would make sense to create a new issue that directly describe the bug if the description of the current PR is too specific to one particular solution?

**Comment by jmooring** [comment]
Your particular use case corresponds to both Example 1 and Example 3 above. Both examples have the same footnote number for each item on the list, which doesn't make any sense to me. This enhancement is considerably more nuanced than a simple implementation of `extension.WithFootnoteIDPrefix` as described in the yuin/goldmark [README](https://github.com/yuin/goldmark?tab=readme-ov-file#footnotes-extension).

Please don't open another issue.

**Comment by shaform** [comment]
Got it. To be specific, previously, with blackfriday, we could achieve example 1 with correct links to each article's footnotes in the list page.

**Comment by shaform** [comment]
> Both examples have the same footnote number for each item on the list, which doesn't make any sense to me.

I use hugo for blogs, and each article feels independent to each other even on list page, so having separate footnotes make sense to me, but I guess people could have different opinions.

**Comment by jmooring** [comment]
> previously, with blackfriday, we could achieve example 1 with correct links

Out of curiosity, was this handled via a front matter field unique to each page, or something else?

**Comment by shaform** [comment]
> > previously, with blackfriday, we could achieve example 1 with correct links
> 
> Out of curiosity, was this handled via a front matter field unique to each page, or something else?

@jmooring Sorry, because it's too long ago, it took some time for me to confirm exactly how it looked like during that time. 

It turns out that for blackfriday, not even `footnoteAnchorPrefix` is needed. That's because in `blackfriday`, if you wrote something like the following:

```markdown

However, it works in blackfriday[^blackfriday-works]

[^blackfriday-works]: It simply works with blackfriday.
```

The link generated by `blackfriday` will be `#fn:blackfriday-works` instead of `#fn:1`. This was at least the case during 2020 before hugo switched to goldmark, but it might not have been the case in 2014. Otherwise the [article](https://discourse.gohugo.io/t/footnote-references-are-duplicated-on-list-pages/116) shouldn't complain about footnote conflicts. ( I don't remember whether this was a default behavior or you need to configure something to make it work like this.)

Therefore, during 2020 before the switch to `goldmark`, if you always use a unique id for each footnote, they will not conflict. However, setting a different prefix for each front matter might have also worked, as it was the solution discussed in the  [article](https://discourse.gohugo.io/t/footnote-references-are-duplicated-on-list-pages/116), and it appears the changes in the related PRs were merged.

Notably, my [original post](https://discourse.gohugo.io/t/goldmark-how-to-configure-the-footnote-anchor-prefix/29639) actually explained about the behavior of blackfriday in my `update 1`, and I did not mention about setting unique prefix in my original post. If you check the edit history, it was [alexandros](https://discourse.gohugo.io/u/alexandros) who suggested that using an unique prefix could be a solution, and edited the title of my post. The suggestion was likely based on the [article](https://discourse.gohugo.io/t/footnote-references-are-duplicated-on-list-pages/116) I had referenced in my post, since the discussions in the referenced article was mainly about setting unique prefixes.

However, since hugo developers cannot change the behavior of `goldmark` to generate links like `#fn:blackfriday-works` , I guess supporting a unique prefix for each article might still be the only viable solution.

**Comment by shaform** [comment]
I've updated the issue description to better reflect the situation.

**Comment by jmooring** [comment]
> [!note]
> This approach does not support pagination.

If each footnote reference within a given page collection is unique, use the approach shown in Example 3 above.

<img width="752" height="544" alt="Image" src="https://github.com/user-attachments/assets/013b25fe-8d1e-4953-ae65-8a566a7cf7ca" />

Try it:

```text
git clone --single-branch -b hugo-github-issue-8045 https://github.com/jmooring/hugo-testing hugo-github-issue-8045
cd hugo-github-issue-8045
hugo server
```

And with Goldmark you can use whatever footnote reference you'd like: `[^42]`, `[^foo]`, etc. When rendered, they will be numbered 1-N going from top down. This is consistent with GitHub Flavored Markdown and the [PHP Markdown Extra](https://michelf.ca/projects/php-markdown/extra/#footnotes) extension, which is probably the most widely adopted specification for footnotes in Markdown.

Note that this was not possible prior to introduction of the [`Page.RenderShortcodes`](https://gohugo.io/methods/page/rendershortcodes) method in [v0.117.0](https://github.com/gohugoio/hugo/releases/tag/v0.117.0).

**Comment by jmooring** [comment]
Closing this out, I've updated the example site to demonstrate two methods for handling footnotes when the content from multiple pages is rendered onto a single page.

```text
git clone --single-branch -b hugo-github-issue-8045 https://github.com/jmooring/hugo-testing hugo-github-issue-8045
cd hugo-github-issue-8045
hugo server
```

On the "Books" section page, footnotes for each page are rendered immediately following that page's content. This corresponds to Example 1 in my previous comments.

On the "Films" sections page, footnotes for all pages are collected and rendered together after all content has been displayed. This corresponds to Example 3 in my previous comments.
